File: src/mqtt/MQTTPublisher.cpp
#include "mqtt/MQTTPublisher.hpp"

XBT_LOG_NEW_DEFAULT_CATEGORY(mqtt_publisher, "MQTT Publisher");

namespace enigma {
namespace mqtt {

namespace sg4 = simgrid::s4u;

MQTTPublisher::MQTTPublisher(const std::string& broker, const std::string& pub_id)
    : broker_name(broker) {
    
    // Generate publisher ID if not provided
    if (pub_id.empty()) {
        auto* host = sg4::this_actor::get_host();
        publisher_id = std::string(host->get_cname()) + "_pub_" + 
                      std::to_string(sg4::this_actor::get_pid());
    } else {
        publisher_id = pub_id;
    }
    
    broker_mbox = sg4::Mailbox::by_name(MQTTBroker::get_broker_mailbox(broker_name));
    
    XBT_DEBUG("MQTT Publisher '%s' initialized (broker: %s)",
              publisher_id.c_str(), broker_name.c_str());
}

void MQTTPublisher::publish(const std::string& topic,
                            const std::string& payload,
                            size_t size,
                            int qos) {
    // Create MQTT message
    auto msg = std::make_shared<MQTTMessage>(topic, payload, size, publisher_id, qos);
    
    // Send to broker
    auto* ctrl_msg = MQTTControlMessage::publish(msg);
    
    XBT_DEBUG("Publishing to topic '%s' (size: %zu bytes, QoS: %d)",
              topic.c_str(), size, qos);
    
    broker_mbox->put(ctrl_msg, sizeof(MQTTControlMessage));
}

void MQTTPublisher::publish(const std::string& topic,
                            const std::string& payload,
                            int qos) {
    // Automatic size calculation
    size_t size = payload.size();
    publish(topic, payload, size, qos);
}

} // namespace mqtt
} // namespace enigma

File: src/python/enigma/comms/mqtt/mqtt_publisher.py
"""
MQTTPublisher: publishes messages to topics via the simulated MQTT broker.
"""
from __future__ import annotations

import simgrid
from .mqtt_broker import MQTTBroker, MQTTMessage, _ControlMessage, _ControlType


class MQTTPublisher:
    """
    Published messages are sent as a :class:`~.mqtt_broker._ControlMessage`
    (type=PUBLISH) to the broker's control mailbox.  The broker then fans
    the message out to all registered subscribers.

    Parameters
    ----------
    broker_name:
        Must match the name used when calling :func:`~.mqtt_broker.start_broker`.
    publisher_id:
        Optional human-readable name; defaults to the current actor's host name.
    """

    def __init__(
        self,
        broker_name: str = "mqtt_broker",
        publisher_id: str = "",
    ) -> None:
        self.broker_name = broker_name
        self._publisher_id = publisher_id or simgrid.this_actor.get_host().name
        self._broker_mbox = simgrid.Mailbox.by_name(
            MQTTBroker.get_broker_mailbox(broker_name)
        )

    # ------------------------------------------------------------------ #
    # Public interface                                                     #
    # ------------------------------------------------------------------ #

    def publish(
        self,
        topic: str,
        payload: str,
        size: int = 0,
        qos: int = 0,
    ) -> None:
        """
        Publish *payload* to *topic*.

        Parameters
        ----------
        topic:
            MQTT topic string, e.g. ``"sensors/temperature"``.
        payload:
            Message body string.
        size:
            Byte size used for SimGrid network simulation.
            Defaults to ``len(payload)`` if 0.
        qos:
            MQTT Quality of Service level (0, 1, or 2).
        """
        byte_size = size if size > 0 else len(payload.encode())
        msg = MQTTMessage(
            topic=topic,
            payload=payload,
            size=byte_size,
            publisher=self._publisher_id,
            qos=qos,
        )
        ctrl = _ControlMessage(
            ctrl_type=_ControlType.PUBLISH,
            topic=topic,
            message=msg,
        )
        # The network transfer cost is counted by the broker mailbox put
        self._broker_mbox.put(ctrl, byte_size)

    @property
    def publisher_id(self) -> str:
        return self._publisher_id
